Close tabs to the right/others. #4124

kcramer opened the pull request:
This PR adds the enhancement from issue #4124.

I added a new function in app shell closeWidgets() and modified closeAll() to use it.  This function is used by the new commands in docmanager-extension to close the widgets.  I added a new helped function to determine the widgets to the right of the chose widget.

I tested the functionality and it works except in the main area.  Right clicking on a tab/widget that is not the active widget does not work correctly, but this is a known issue.  It is discussed and already addressed in #4529.  telamonian's solution worked correctly when I tested it.  When that functionality is merged, my changes just need to be updated from using app.shell.widget to his new contextMenuWidget() function.
 
The context menu does work correctly when right clicking on a tab from the left sidebars Tabs pane.

afshin commented:
Hi! Thanks for working on this! I have left some comments in a review.

kcramer commented on packages/application/src/shell.ts:
```diff
@@ -488,7 +488,14 @@ class ApplicationShell extends Widget {
+  /**
+   * Close widgets in the given array.
+   */
+  closeWidgets(widgets: Array<Widget>): void {
```
I originally had all the code, even the selection of which widgets to close, in app shell.  I can move the closeWidgets function to docmanager-extension.

Should I also move closeAll() then?  If not, I could update closeAll() to use forEach instead of each.

kcramer commented on packages/docmanager-extension/src/index.ts:
```diff
@@ -206,6 +229,42 @@ function addCommands(app: JupyterLab, docManager: IDocumentManager, palette: ICo
+  commands.addCommand(CommandIDs.closeOtherTabs, {
+    label: () => `Close Other Tabs`,
+    isEnabled: () => {
+      const widgets = toArray(app.shell.widgets('main'));
```
I wanted to check that at least two widgets exists, otherwise there is no need to show the Close Others option since there are no other tabs.  I had originally used the iterator thinking it was more efficient but I didn't like the way it looked so I decided to use code which was clearer on the intent even if it wasn't as efficient.

I can use the iterator if you think that is best.  I just need to call next() twice to ensure there are two widgets.

kcramer commented on packages/docmanager-extension/src/index.ts:
```diff
@@ -181,6 +195,15 @@ function addCommands(app: JupyterLab, docManager: IDocumentManager, palette: ICo
   };
+  const widgetsRightOf = (widget: Widget): Array<Widget> => {
+    const { id } = widget;
+    const widgets = toArray(app.shell.widgets('main'));
```
I'm not sure of the difference on the first comment.   In my testing, it worked properly but I may not know all the test cases.

kcramer commented:
I made the changes you requested.  I cached shell at the top of the addCommands() function and then updated all references to app.shell in that function.

The only unaddressed issue was on widgets('main') providing all the widgets in the main area's dock panel and that may not be what is desired.  In my use of Jupyterlab I haven't seen where it would possibly diverge or how to access the smaller set.  If I'm missing something please point me in the right direction and how to access the subset.

afshin commented:
@kcramer Hm, actually @sccolbert refactored the dock panel so that all the child widgets are siblings, irrespective of whether they are grouped together.

So the only way I can think of for you to achieve what you want to do consists of the following: the `ApplicationShell` has a `saveLayout()` method that returns an object that describes your current layout:

``` typescript
const { mainArea } = shell.saveLayout();

// Bail if not in multiple document mode.
if (mainArea.mode !== 'multiple-document') {
  return;
}

// Recurse through the mainArea.dock data structure and find the tab area
// that contains the widget you care about so that you can find all the
// widgets that are in the same tab area and come after your widget.
```
The layout configuration data structure that you get is defined in these two files (not just in the highlighted lines, but those are a good place to start):

1. https://github.com/jupyterlab/jupyterlab/blob/master/packages/application/src/shell.ts#L734-L761
1. https://github.com/phosphorjs/phosphor/blob/master/packages/widgets/src/docklayout.ts#L1247-L1259

afshin commented:
For example, consider this layout:
<img width="1199" alt="A layout with two tab areas" src="https://user-images.githubusercontent.com/159529/41783492-dfabb9f8-7634-11e8-9391-b2a2749b8d31.png">

You can see that there is a `split-area` that contains two `tab-area` children, the first of which has two child widgets and the second has three widgets.

afshin commented on packages/docmanager-extension/src/index.ts:
```diff
@@ -181,29 +195,73 @@ function addCommands(app: JupyterLab, docManager: IDocumentManager, palette: ICo
+    const widgets = toArray(shell.widgets('main'));
+    const index = widgets.findIndex( widget => widget.id === id );
+    if (index < 0) {
+      return Array<Widget>();
```
Could just be: `return [];` because of type inference.

afshin commented on packages/application/src/shell.ts:
```diff
@@ -488,7 +488,7 @@ class ApplicationShell extends Widget {
     // before removing them because removing them while iterating through them
     // modifies the underlying data of the iterator.
-    each(toArray(this._dockPanel.widgets()), widget => { widget.close(); });
+    toArray(this._dockPanel.widgets()).forEach(widget => { widget.close(); });
```
Good catch, cheers!

afshin commented on packages/docmanager-extension/src/index.ts:
```diff
@@ -181,29 +195,73 @@ function addCommands(app: JupyterLab, docManager: IDocumentManager, palette: ICo
+  const widgetsRightOf = (widget: Widget): Array<Widget> => {
+    const { id } = widget;
+    const widgets = toArray(shell.widgets('main'));
+    const index = widgets.findIndex( widget => widget.id === id );
```
Please format this without spaces inside the parens:
`const index = widgets.findIndex(widget => widget.id === id);`
We only use spaces in object literals: `{ foo: 'bar' }`

kcramer commented:
With the prettier changes plus all the commits since I submitted the PR, I went ahead and just rebased to clean it up.  I fixed the issue with Close Tabs to Right.  Thank you for the earlier explanations.

It only closes tabs in the same tab area now.  I added new helper functions _findTab_ and _tabAreaFor_ which start at line 194 of docmanager-extension/src/index.ts.  I've re-tested it and ran the tests and everything passes.

I think the only remaining question is whether Close Others should also be limited by the current tab group or apply to all other open tabs.  I can see arguments for either approach.

afshin commented:
Cheers @kcramer! I think the current behavior is reasonable, so I think we should merge this now and see how people like it.

kcramer commented:
Thanks @afshin!  I can easily change it, if people prefer it to work differently.
